sections/server.tex

```diff
@@ -43,3 +43,77 @@ \section{QuizFight server}
+one or more required fields are missing, the server notifies the client using an
+error response.
+
+Using JSON arises a problem at client side. Java is a strongly typed language, so we
+cannot freely manipulate objects as in JavaScript. We need a way to cast a JSON 
+object into a Java (POJO) one. We do that using Gson, a Google's library for 
+serializing and deserializing JSON objects. 
+
+Following the object-oriented programming paradigm we design a scalable and easily 
+manageable architecture for REST calls. Figure~\ref{fig:retrofit} shows a partial
+class diagram for that. Only two concrete API calls are shown; for the others it 
+is exactly the same.
+\begin{figure}[h]
+	\centering
+	\includegraphics[width=0.9\linewidth]{Retrofit}
+	\caption{High-level class diagram for API calls}
+	\label{fig:retrofit}
+\end{figure}
+
+\texttt{RetrofitHelper} is the actual entry point. Its method \texttt{getAPIService()}
+configures every parameter required, such as the base URL, the timeout, and the 
+library for dealing with JSON. Then, via reflection, it creates on the fly a 
+concrete \texttt{EndpointInterface} instance. The latter actually contains the
+methods corresponding to the API calls. Retrofit allows us to specify HTTP
+parameters by using Java-based annotations, such as \texttt{@Headers} for the
+headers, and \texttt{@PUT("endpoint")} for a PUT call. 
+
+Those methods return a generic \texttt{Call<T>} object. The type parameter 
+represents the POJO object returned by the server. In this sense Retrofit is a
+type-safe REST library. The \texttt{EndpointInterface} concrete instance is then
+used by an abstract class, \texttt{APICaller}, which basically implements a mix
+of the template method and command design patterns. Each concrete class extending
+\texttt{APICall} should have a field representing the object to be sent, 
+as a parameter, to the server. Then it implements the abstract method 
+\texttt{getActualMethod()} in order to make the right call with the right 
+parameters. It is invoked by the \texttt{APICall}'s \texttt{call()} method,
+which also enqueues the actual execution to be performed in an asynchronous way. 
+
+Then, whenever a Java class wishes to communicate with the server, it must 
+instantiate the corresponding API class, give a parameter to it and provide
+two callbacks, \texttt{onResponse()} and \texttt{onFailure()}, for handling the
+corresponding situations. Listing~\ref{lst:servercall} shows an example.
+
+\begin{lstlisting}[language=Java, caption={Server call example}, label={lst:servercall}]
+new AddToken(new User(
+	Games.Players.getCurrentPlayer(client).getDisplayName(),
+	token,
+	Secure.getString(getContentResolver(), Secure.ANDROID_ID)
+)).call(new Callback<ResponseBody>() {
+	@Override
+	public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {}
+	
+	@Override
+	public void onFailure(Call<ResponseBody> call, Throwable t) {}
+});
```

Comment:
In a recent issue, this function `call(@NonNull Callback<T> callback)` becomes `call(@NonNull Callback<T> callback, QuizFightApplication application)`, it requires also the application parameter